src/oid4vci/http_client.rs

```diff
@@ -0,0 +1,151 @@
+
+    #[error("failed to parse header")]
+    HeaderParse,
+
+    #[error("failed to parse header key: {key}")]
+    HeaderKeyParse { key: String },
+
+    #[error("failed to parse header value: {value}")]
+    HeaderValueParse { value: String },
+
+    #[error("failed to parse header entry: ({key}, {value})")]
+    HeaderEntryParse { key: String, value: String },
+
+    #[error("other error: {error}")]
+    Other { error: String },
+}
+
+impl From<String> for HttpClientError {
+    fn from(value: String) -> Self {
+        Self::Other { error: value }
+    }
+}
+
+#[derive(uniffi::Record, Clone, Debug)]
+pub struct HttpRequest {
+    pub url: String,
+    pub method: String,
+    pub headers: HashMap<String, String>,
+    pub body: Vec<u8>,
+}
+
+impl TryFrom<IHttpRequest> for HttpRequest {
+    type Error = HttpClientError;
+
+    fn try_from(req: IHttpRequest) -> Result<Self, Self::Error> {
+        Ok(Self {
+            url: req.uri().to_string(),
+            method: req.method().to_string(),
+            headers: headermap_to_hashmap(req.headers())?,
+            body: req.body().clone(),
+        })
+    }
+}
+
+impl TryInto<IHttpRequest> for HttpRequest {
+    type Error = HttpClientError;
+
+    fn try_into(self) -> Result<IHttpRequest, Self::Error> {
+        let mut request = Request::builder()
+            .method(Method::from_str(&self.method).map_err(|_| HttpClientError::MethodParse)?)
+            .uri(Uri::from_str(&self.url).map_err(|_| HttpClientError::UrlParse)?);
+
+        for (k, v) in self.headers {
+            request = request.header(k, v);
+        }
+
+        request
+            .body(self.body)
+            .map_err(|_| HttpClientError::RequestBuilder)
+    }
+}
+
+#[derive(uniffi::Record, Clone, Debug)]
+pub struct HttpResponse {
+    pub status_code: u16,
+    pub headers: HashMap<String, String>,
+    pub body: Vec<u8>,
+}
+
+impl TryFrom<IHttpResponse> for HttpResponse {
+    type Error = HttpClientError;
+
+    fn try_from(res: IHttpResponse) -> Result<Self, Self::Error> {
+        Ok(Self {
+            status_code: res.status().as_u16(),
+            headers: headermap_to_hashmap(res.headers())?,
+            body: res.body().clone(),
+        })
+    }
+}
+
+impl TryInto<IHttpResponse> for HttpResponse {
+    type Error = HttpClientError;
+
+    fn try_into(self) -> Result<IHttpResponse, Self::Error> {
+        let mut response = Response::builder().status(
+            StatusCode::from_u16(self.status_code)
+                .map_err(|_| "failed to parse status code".to_string())
+                .map_err(HttpClientError::from)?,
+        );
+
+        for (k, v) in self.headers {
+            response = response.header(k, v);
+        }
+
+        response
+            .body(self.body)
+            .map_err(|_| HttpClientError::ResponseBuilder)
+    }
+}
+
+// TODO figure out how to use async_trait to have an async version of HttpClient
+#[uniffi::export(with_foreign)]
+pub trait HttpClient: Send + Sync {
```

Comment:
IMO let's not worry about doing HTTP calls from the native code, and just use `oid4vci::openidconnect::reqwest::Client`

Reply:
I went with a trait because it'll be useful to let the library user potentially configure different parameters for an HTTP client that are not specific to the OID4VCI spec, like extra headers, timeouts, certificates, etc.
But we can definitely have `oid4vci::openidconnect::reqwest::Client` be the default.

Reply:
Agreed. We could offer a default with `reqwest` but let's do that in a separate PR if we find the need for it. It doesn't seem hard to implement the HTTP client in Swift/Kotlin and offers better debugging capabilities.